### PR TITLE
Stop the Aspire CLI and own its RPC connections during extension deactivation

### PR DESCRIPTION
**Pull request by adamint**
## Description

Split out of #19124, which combined notification presentation with extension shutdown/lifecycle hardening. This branch now contains only the shutdown work.

Extension deactivation was fire-and-forget: `deactivate()` returned `void`, so VS Code did not wait for CLI stop requests and closing a window could leave `aspire run` processes alive. RPC connections still inside the debug-session handshake were not owned by the server, and late status updates could repaint progress after transport disposal.

The extension now has one bounded shutdown path:

- `deactivate()` returns `AspireExtensionContext.deactivate()` so VS Code can await it.
- The context stops every live or newly registered session, waits up to five seconds, then force-cleans remaining owned processes and transports.
- Debug sessions reuse in-flight CLI stop requests, wait for a late RPC client when needed, and prevent a CLI from spawning after shutdown starts.
- Long-lived POSIX CLI launches own a process group; Windows cleanup avoids targeting an exited/reused leader PID.
- `AspireRpcServer` owns connections from creation, including pending handshakes, and rejects connections added after disposal.
- RPC clients and interaction services dispose idempotently and ignore late status updates.

The E2E now starts a real AppHost through the generated VSIX and repo CLI, records the owned CLI and AppHost process IDs, invokes the process-owner cleanup path while keeping the extension host alive, and verifies both processes and the AppHost/debug-session state stop.

## Verification

```console
corepack yarn --silent compile-tests
corepack yarn --silent compile-e2e
corepack yarn --silent unit-test --grep "AspireExtensionContext|AspireDebugSession tests|CLI process termination|spawnCliProcess tests|AspireRpcServer|InteractionService endpoints"

  122 passing
  1 pending
```

Additional exact-head proof:

- Pending RPC/handshake boundary: 9 passing
- Focused `edgeCases.e2e.test.js`: 5/5 passing against the generated VSIX and repo CLI
- Exported `deactivate()` wiring: focused unit coverage verifies VS Code receives the context shutdown promise
- Real process-owner cleanup E2E: 1/1 passing; owned CLI and AppHost processes terminated and extension state returned to no running AppHost/debug sessions
- Windows process-tree behavior remains unit-tested; the real process-owner cleanup proof ran on macOS

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
  - [x] No

### CODE REVIEW

File: extension/src/utils/workspace.ts [hunk 1]
```diff
@@ -54,19 +54,28 @@ export function isFolderOpenInWorkspace(folderPath: string): boolean {
...
 
 export function getRelativePathToWorkspace(filePath: string): string {
-    if (!isWorkspaceOpen(false)) {
-        return filePath;
-    }
-
     const uri = vscode.Uri.file(filePath);
     const workspaceFolder = vscode.workspace.getWorkspaceFolder(uri);
+    if (!workspaceFolder) {
+        return path.basename(filePath);
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/utils/workspace.ts]: Fixed in `f7e616a7`: the rejection moved ahead of the `getWorkspaceFolder` early return, and foreign paths reduce via `path.win32.basename`, which splits on both separators. The old test stubbed a folder, so it never ran this path; `reduces absolute paths from either platform to a bare file name` now stubs `undefined` and covers `C:\...`, UNC and `/home/...`.

File: extension/src/AspireExtensionContext.ts [hunk 2]
```diff
@@ -94,14 +101,101 @@ export class AspireExtensionContext implements vscode.Disposable {
...
+        }
+
+        if (this._shutdownPromise) {
+            return this._shutdownPromise;
+        }
+
+        this._isShuttingDown = true;
+        // Schedule the async work after storing the shared promise so a reentrant dispose/deactivate
+        // call cannot begin synchronous teardown between the stop request and the first await.
+        this._shutdownPromise = Promise.resolve().then(() => this._deactivateCore());
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts:140]: Fixed in `938888da`: `_waitForCliStopRequests` now re-snapshots the session array between awaits until no new session appears, so one registered mid-await is still asked to stop. Covered by `deactivation stops a debug session registered while an earlier stop is in flight`.

File: extension/src/AspireExtensionContext.ts [hunk 3]
```diff
@@ -94,14 +101,101 @@ export class AspireExtensionContext implements vscode.Disposable {
...
+            }),
+        ]);
+
+        if (timeout) {
+            clearTimeout(timeout);
+        }
+
+        if (outcome.timedOut) {
+            extensionLogOutputChannel.warn(`Timed out after ${AspireExtensionContext._cliStopTimeoutMs}ms waiting for Aspire CLI stop requests; continuing extension teardown.`);
+            return;
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts]: Correct — fixed in `938888da`. `spawnAspireCommand` now retains the `ChildProcess` and spawns `aspire run` with `createProcessGroup: true`, and the new `terminateCliProcessTree()` signals it after the cooperative stop settles or times out. Covered by three new tests including the never-settles case.

File: extension/src/AspireExtensionContext.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts:140]: Fixed across `938888da` and `67519d5f33`. The stop drain now re-scans the session array between awaits, so a descriptor or RPC-triggered session landing mid-shutdown is still asked to stop. `addAspireDebugSession` also refuses and disposes once `_isDisposed` is set — that was the half your comment flagged where a post-`_disposeCore` arrival is never disposed at all.

File: extension/src/AspireExtensionContext.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts]: Correct on both counts, fixed in `938888da`/`67519d5f33`. `spawnAspireCommand` now retains the child, spawns it as a process-group leader, and escalates to `terminateCliProcess` when the cooperative stop does not confirm — timeout and `PendingResponseRejected` included. It also no longer spawns at all if disposal won that race during the CLI-path await.

File: extension/src/debugger/AspireDebugSession.ts [hunk 4]
```diff
@@ -412,6 +471,8 @@ export class AspireDebugSession implements vscode.DebugAdapter {
           vscode.window.showErrorMessage(processExceptionOccurred(error.message, commandLabel));
         },
         exitCallback: (code) => {
+          // The CLI came down on its own, so the escalation timer has nothing left to signal.
+          this.cancelScheduledCliProcessTermination();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts]: Right, and `terminateCliProcess` already reaps a managed group whose leader exited — it just was never invoked on that path. In 512c887ad1 the exit callback now calls `terminateCliProcessTree()` (which cancels the timer itself) and the exited-leader early return is gone. Done synchronously there because a released PID can be recycled as another group's id.

File: extension/src/AspireExtensionContext.ts [hunk 5]
```diff
@@ -94,14 +101,145 @@ export class AspireExtensionContext implements vscode.Disposable {
...
+                break;
+            }
+        }
+
+        // A cooperative stop that resolved, rejected or timed out proves only what happened to the
+        // RPC request; the CLI process can still be running. Signal any that are, so deactivation
+        // cannot leave an AppHost and its resource processes orphaned.
+        for (const session of this._aspireDebugSessions) {
+            try {
+                session.terminateCliProcessTree();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts]: Confirmed — the force-kill timer is `unref()`d and `_deactivateCore` resolves as soon as the sweep returns. 512c887ad1 takes your first option: `terminateCliProcess` gains `force`, which skips SIGTERM and the timer and signals the group immediately, and the deactivation sweep passes it. The in-session 10s escalation keeps the graceful path, since the host is still alive there.

File: extension/src/AspireExtensionContext.ts [hunk 6]
```diff
@@ -94,14 +112,149 @@ export class AspireExtensionContext implements vscode.Disposable {
...
+        // A cooperative stop that resolved, rejected or timed out proves only what happened to the
+        // RPC request; the CLI process can still be running. Signal any that are, so deactivation
+        // cannot leave an AppHost and its resource processes orphaned.
+        for (const session of this._aspireDebugSessions) {
+            try {
+                // Force rather than signal-and-schedule: `terminateCliProcess` escalates to a hard
+                // kill on an `unref`'d timer, and `_deactivateCore` resolves immediately after this
+                // sweep, so the extension host can exit before that timer fires. The cooperative
+                // deadline above was this CLI's grace period; there is no second one.
+                session.terminateCliProcessTree({ force: true });
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts:188]: Confirmed: with the leader already exited, terminateCliProcess takes the else branch and returns, and on Windows processGroupPid is always undefined, so the force path never runs and no taskkill /t /f is issued. Killing the tree before the exit RPC would defeat the cooperative stop, so this needs Windows descendant ownership (job object or captured PIDs). Leaving open for that work rather than patching it here.

File: extension/src/debugger/AspireDebugSession.ts [hunk 7]
```diff
@@ -429,16 +508,27 @@ export class AspireDebugSession implements vscode.DebugAdapter {
...
+        void this.requestCliStopForExtensionShutdown().catch((err) => {
           extensionLogOutputChannel.info(`stopCli failed (connection may already be closed): ${err}`);
         });
         extensionLogOutputChannel.info(`Requested Aspire CLI exit with args: ${args.join(' ')}`);
+        // `stopCli` is cooperative and cannot be the only stop mechanism: it resolves without
+        // effect when the transport is already closed, and never settles when the CLI has stopped
+        // servicing the connection. Escalate to signalling the process group once the CLI has had
+        // a chance to exit on its own, so a CLI that ignores the request cannot outlive the
+        // session and keep the AppHost and its resource processes alive.
+        this.scheduleCliProcessTermination();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts:626]: Verified: removeAspireDebugSession is the first disposable (line 131) and the escalation is scheduled by a later one (line 520) on an unref()d 10s timer, so dispose() untracks the session while a termination is still pending. Keeping it in _aspireDebugSessions would make id lookups return a dead session, so this needs the extension-level owner you describe. Leaving open.

File: extension/src/AspireExtensionContext.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/AspireExtensionContext.ts:188]: Fixed in 64bcad3860. `terminateCliProcess` now still launches `taskkill.exe /pid <pid> /t` for an already-exited Windows leader, adding `/f` for force shutdown; covered by win32 platform-stub tests.

File: extension/src/debugger/AspireDebugSession.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts:626]: Fixed in 64bcad3860. Disposed sessions with pending CLI termination now stay tracked as shutdown owners but are hidden from active-session lookups, so deactivation can still force-drain them.

File: extension/src/debugger/languages/cli.ts [hunk 8]
```diff
@@ -165,6 +165,16 @@ export function terminateCliProcess(childProcess: ChildProcessWithoutNullStreams
         return;
     }
 
+    if (options?.force) {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/languages/cli.ts:178]: Fixed in 64bcad3860. `terminateCliProcess` now still launches `taskkill.exe /pid <pid> /t` for an already-exited Windows leader, adding `/f` for force shutdown; covered by win32 platform-stub tests.

File: extension/src/debugger/AspireDebugSession.ts [hunk 9]
```diff
@@ -135,6 +149,73 @@ export class AspireDebugSession implements vscode.DebugAdapter {
...
+   * Signals the `aspire` CLI process tree.
+   *
+   * The cooperative `stopCli` RPC resolving proves only that the request was accepted, and on a
+   * closed transport it resolves having done nothing at all. Neither outcome terminates the CLI,
+   * so the process it owns — the AppHost and every resource process beneath it — has to be
+   * signalled directly whenever the cooperative path did not finish the job. The leader may already
+   * have exited by then, but that does not prove its descendants exited too.
+   */
+  terminateCliProcessTree(options?: { force?: boolean }): void {
+    this.cancelScheduledCliProcessTermination();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts:192]: Fixed in 380cdb385c. terminateCliProcessTree now returns once _cliProcessTreeTerminationAttempted is set, with unit coverage for a force call followed by a second call.

[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts:192]: Follow-up pushed in 1f6727825b. The exit callback now uses the force path, and `terminateCliProcessTree` is idempotent before calling `terminateCliProcess`, so a deactivation force sweep cannot be followed by a second taskkill for the same recorded PID. Regression tests cover the force-then-exit sequence.

File: extension/src/debugger/AspireDebugSession.ts [hunk 10]
```diff
@@ -412,6 +508,12 @@ export class AspireDebugSession implements vscode.DebugAdapter {
           vscode.window.showErrorMessage(processExceptionOccurred(error.message, commandLabel));
         },
         exitCallback: (code) => {
+          // The leader came down on its own, so the escalation timer has nothing left to signal —
+          // but a detached leader's descendants (the AppHost and every resource process beneath it)
+          // can outlive it, and this is the last moment they can be collected safely: once the
+          // leader's PID is released the operating system may recycle it as another group's id, and
+          // a later negative-PID signal would land on something unrelated.
+          this.terminateCliProcessTree();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts]: Fixed in 1f6727825b by making the exit cleanup call `terminateCliProcessTree({ force: true })`. The pushed file at head has the force call immediately before `recordAppHostProcessExit`, so the Windows taskkill path still runs for an already-exited leader before the idempotence guard blocks repeats.

File: extension/src/debugger/AspireDebugSession.ts [hunk 11]
```diff
@@ -412,6 +517,13 @@ export class AspireDebugSession implements vscode.DebugAdapter {
           vscode.window.showErrorMessage(processExceptionOccurred(error.message, commandLabel));
         },
         exitCallback: (code) => {
+          // A detached POSIX leader's descendants can keep the process group alive after the
+          // leader exits, and the group id can be reused later, so collect that group immediately.
+          // Windows taskkill needs the target PID to still identify a live process tree; after the
+          // close event the CLI PID may already be reusable, so do not taskkill from this path.
+          if (process.platform !== 'win32') {
+            this.terminateCliProcessTree({ force: true });
+          }
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts:585]: Confirmed and fixed in 73b6048b3e.

You were right that the skip did not hold. `_cliProcess` is never cleared, so `dispose()` re-ran the CLI disposable, `scheduleCliProcessTermination()` saw no timer and no prior attempt, and it scheduled the forced sweep of the just-released PID for ten seconds later — the exact hazard the comment two lines up claims to avoid.

The Windows branch now calls `abandonCliProcessTree()`, which cancels any pending escalation and marks the recorded PID as spent so the disposable cannot schedule another. A CLI still running at extension shutdown is unaffected, since that path never reaches the exit callback.

Also took your point about the test: the old one asserted only at exit time, so it passed while the delayed sweep was live. The new test advances 30s past the 10s grace period. Removing the fix fails that test and only that test (1488 passing, 1 failing).

File: extension/src/debugger/AspireDebugSession.ts [hunk 12]
```diff
@@ -412,6 +517,12 @@ export class AspireDebugSession implements vscode.DebugAdapter {
           vscode.window.showErrorMessage(processExceptionOccurred(error.message, commandLabel));
         },
         exitCallback: (code) => {
+          // The leader came down on its own, so the escalation timer has nothing left to signal —
+          // but a detached leader's descendants (the AppHost and every resource process beneath it)
+          // can outlive it, and this is the last moment they can be collected safely: once the
+          // leader's PID is released the operating system may recycle it, and a later process-tree
+          // signal could land on something unrelated.
+          this.terminateCliProcessTree({ force: true });
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts]: Same root cause as the line 526 thread, fixed in 73b6048b3e — the recorded PID is now retired when the Windows leader exits, so nothing schedules or issues a taskkill against it afterwards.

File: extension/src/debugger/AspireDebugSession.ts [hunk 13]
```diff
@@ -396,9 +488,20 @@ export class AspireDebugSession implements vscode.DebugAdapter {
       return partial;
     };
 
-    spawnCliProcess(
+    const cliPath = await this._terminalProvider.getAspireCliExecutablePath();
+    if (this._disposed) {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/AspireDebugSession.ts]: Fixed, and verified at `9689a52a39`.

Shutdown is now latched on the session before the sweep runs, rather than the launch path relying on `_disposed` alone. `requestCliStopForExtensionShutdown()` sets the latch as its first action, so it is already set for every session the sweep visits:

```ts
requestCliStopForExtensionShutdown(): Promise<void> {
    this._extensionShutdownRequested = true;
    if (!this._rpcClient) {
      return Promise.resolve();
    }

    this._cliStopPromise ??= this._rpcClient.stopCli();
    return this._cliStopPromise;
}
```

The launch path re-checks both flags after the asynchronous CLI-path resolution, which is the continuation point the finding identified:

```ts
const cliPath = await this._terminalProvider.getAspireCliExecutablePath();
if (this._disposed || this._extensionShutdownRequested) {
  // Resolving the CLI path is asynchronous, so extension deactivation can complete between the
  // launch request and this point. Spawning now would produce an `aspire run` that no teardown
  // path is left to stop — and because it is spawned detached as a process-group leader, it
  // would not even die with the extension host.
  extensionLogOutputChannel.info(`Skipping Aspire CLI launch for disposed or shutting-down debug session ${this.debugSessionId}.`);
  disposable.dispose();
  return;
}
```

`AspireExtensionContext._waitForCliStopRequests()` reaches every non-disposed session through `requestCliStopForExtensionShutdown()`, so by the time the final force sweep completes, no session can still be in a state where a pending `getAspireCliExecutablePath()` continuation would spawn. A launch that was already in flight when deactivation began now returns without spawning instead of creating a detached process after the sweep.

Covered by `a launch that resolves the CLI path after extension shutdown was requested does not spawn an orphan CLI`, which resolves the CLI path only after `requestCliStopForExtensionShutdown()` has been awaited and asserts no process is spawned — deliberately distinct from the sibling case that covers the `_disposed` route. Both pass at this SHA (`2 passing`).

File: extension/src/debugger/languages/cli.ts [hunk 14]
```diff
@@ -161,7 +162,30 @@ export function terminateCliProcess(childProcess: ChildProcessWithoutNullStreams
...
+
+        if (!isWindows) {
+            return;
+        }
+
+        // Windows does not tie child lifetimes to the parent process. An exited CLI leader can
+        // still have an AppHost/resource tree underneath its recorded PID, so sweep it with
+        // taskkill during forceful shutdown instead of treating the leader's exit as proof that
+        // teardown completed. Non-force callers keep the historical no-op behavior because a short
+        // helper CLI can legitimately exit before its close handler observes it.
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/debugger/languages/cli.ts]: Checked this against `cli.ts` at the current head and the post-exit `taskkill` race is already closed on every Windows path.

Tracing them individually:

- the `exited` early-return fires before any termination work begins;
- both the `options?.force` branch and the graceful signal are nested inside `if (!exited)`, so neither can issue a kill after exit;
- the escalation timer re-checks `exitCode` itself when it fires, which covers the window between scheduling and firing;
- `onExit`'s call to `forceTermination()` is POSIX-gated, so it is not reachable on Windows at all.

No code change needed.

File: extension/src/extension.ts [hunk 15]
```diff
@@ -423,8 +423,8 @@ export async function activate(context: vscode.ExtensionContext) {
   return Object.freeze(api);
 }
 
-export function deactivate() {
-  aspireExtensionContext.dispose();
+export function deactivate(): Promise<void> {
```
Comment (ellahathaway) [inline extension/src/extension.ts:427]: `AspireExtensionContext.deactivate()` is covered, but I don't think this export wiring is. The context tests call `deactivate()` directly, and the E2E calls `stopOwnedDebugSessionProcesses`, so changing this back to the old fire-and-forget `dispose()` would leave the focused suite green. Consider adding a focused test that stubs the context and asserts this function returns the same shutdown promise.

File: extension/src/test-e2e/edgeCases.e2e.test.ts [hunk 16]
```diff
@@ -139,4 +145,34 @@ suite('Aspire extension edge case E2E', function () {
...
+        await waitForRunningAppHost(180000);
+
+        const processInfoStatus = await executeE2eControlCommand({ name: 'getDebugSessionProcessInfo', appHostPath });
+        const processInfo = processInfoStatus.result as DebugSessionProcessInfo | undefined;
+        assert.ok(processInfo?.cliPid, `Expected the E2E bridge to report the owned Aspire CLI pid: ${JSON.stringify(processInfoStatus)}`);
+        assert.ok(processInfo?.appHostPid, `Expected the E2E bridge to report the owned AppHost pid: ${JSON.stringify(processInfoStatus)}`);
+        assert.ok(isProcessAlive(processInfo.cliPid), `Expected the Aspire CLI process ${processInfo.cliPid} to be running before deactivation.`);
+        assert.ok(isProcessAlive(processInfo.appHostPid), `Expected the AppHost process ${processInfo.appHostPid} to be running before deactivation.`);
+
+        // This exercises the process-owner cleanup methods with real CLI and AppHost processes while
```
Comment (ellahathaway) [inline extension/src/test-e2e/edgeCases.e2e.test.ts:168]: The description no longer matches this test. It still says the E2E performs a literal workbench reload and verifies RPC closure/reactivation, while this path deliberately keeps the extension host alive and calls the cleanup bridge directly. Consider updating the verification section to describe the process-owner coverage that remains at HEAD.

**Review by ellahathaway — APPROVED** [review]
I reviewed the shutdown flow end to end, including cooperative CLI stop, timeout/force termination, RPC ownership, and disposal ordering. I didn't find a new correctness issue beyond the concerns already addressed in resolved threads. I left two nonblocking coverage/description notes.

File: extension/src/extension.ts (hunk 15 above)
Comment (adamint) [reply extension/src/extension.ts:427]: Fixed in 8039a1233d. Added focused entrypoint coverage that stubs `AspireExtensionContext.prototype.deactivate` and asserts the exported `deactivate()` returns the exact shutdown promise.

File: extension/src/test-e2e/edgeCases.e2e.test.ts (hunk 16 above)
Comment (adamint) [reply extension/src/test-e2e/edgeCases.e2e.test.ts:168]: Updated the PR description. It now describes the live process-owner cleanup E2E at HEAD and no longer claims a literal workbench reload or RPC reactivation.
